src/renderer/base/renderer.cpp

```diff
@@ -543,16 +531,16 @@ void Renderer::UpdateSoftFont(const gsl::span<const uint16_t> bitPattern, const
     const auto softFontCharCount = cellSize.cy ? bitPattern.size() / cellSize.cy : 0;
     _lastSoftFontChar = _firstSoftFontChar + softFontCharCount - 1;
 
-    for (const auto pEngine : _rgpEngines)
+    foreachEngine(pEngine)
     {
         LOG_IF_FAILED(pEngine->UpdateSoftFont(bitPattern, cellSize, centeringHint));
     }
     TriggerRedrawAll();
 }
 
-bool Renderer::s_IsSoftFontChar(const std::wstring_view& v, const size_t firstSoftFontChar, const size_t lastSoftFontChar)
+bool Renderer::_isSoftFontChar(const std::wstring_view& v) const noexcept
 {
-    return v.size() == 1 && v.front() >= firstSoftFontChar && v.front() <= lastSoftFontChar;
+    return v.size() == 1 && v[0] >= _firstSoftFontChar && v[0] <= _lastSoftFontChar;

```

Comment:
If we make this function non-static we can access the members directly and don't need to pass them as arguments. Also `.first()` does bounds-checking.

Reply:
I think the audit mode analyzer doesn't like `v[0]`...

Reply:
Note that the size check in the `front()` call is optimized out when there is already a size check right next to it like there is here. And this function was deliberately changed to static because it wasn't always inlining as a non-static method. If you can get it inlining without the static that would be great.

Reply:
I've taken your code as it is on the main branch and modified `s_IsSoftFontChar` to be non-static function. It still properly inlined the function:
```asm
        // Determine whether we're using a soft font.
        auto usingSoftFont = s_IsSoftFontChar(it->Chars());
00007FF649C547EA  cmp         qword ptr [r15+8],1  
00007FF649C547EF  jne         Microsoft::Console::Render::Renderer::_PaintBufferOutputHelper+12Eh (07FF649C5480Eh)  
00007FF649C547F1  mov         rax,qword ptr [r15]  
00007FF649C547F4  movzx       ecx,word ptr [rax]  
00007FF649C547F7  cmp         rcx,qword ptr [r14+0B0h]  
00007FF649C547FE  jb          Microsoft::Console::Render::Renderer::_PaintBufferOutputHelper+12Eh (07FF649C5480Eh)  
00007FF649C54800  cmp         rcx,qword ptr [r14+0B8h]  
00007FF649C54807  ja          Microsoft::Console::Render::Renderer::_PaintBufferOutputHelper+12Eh (07FF649C5480Eh)  
00007FF649C54809  mov         sil,1  
00007FF649C5480C  jmp         Microsoft::Console::Render::Renderer::_PaintBufferOutputHelper+131h (07FF649C54811h)  
00007FF649C5480E  xor         sil,sil  
```

So I believe we should be able to remove the `static`. 🙂

Reply:
Oh and btw: I'm going to severly shrink this PR today as I feel like it does _way_ too many things at once.

Reply:
It was just the second usage of `s_IsSoftFontChar` that wasn't inlined when non-static, and since that was inside the main loop, that's the one we were worried about. I don't think I've got the latest version of VS, though, so maybe the optimizer has improved since then.

Reply:
> I don't think I've got the latest version of VS, though, so maybe the optimizer has improved since then.

Oh that's a good point! I was using VS 2022! And I didn't check the call in the inner loop...
I've compiled the project with 16.10.4 now and tested this in the inner loop. And indeed! You're right! It doesn't inline the call on VS 2019! 😯🙂 I'll add a note to the function stating why it's necessary. But it might take a while until I can clean up this PR, which is why I marked it as Draft now.

I certainly feel validated in my opinion now, to not trust a C++ compiler a single inch. 😄